java/dagger/internal/codegen/SimpleMethodBindingExpression.java

```diff
@@ -176,13 +176,16 @@ private Expression injectMembers(CodeBlock instance) {
   }
 
   private Optional<CodeBlock> moduleReference(ClassName requestingClass) {
-    return provisionBinding.requiresModuleInstance()
-        ? provisionBinding
-            .contributingModule()
-            .map(Element::asType)
-            .map(ComponentRequirement::forModule)
-            .map(module -> componentRequirementExpressions.getExpression(module, requestingClass))
-        : Optional.empty();
+    if (provisionBinding.requiresModuleInstance(compilerOptions)) {
+      return provisionBinding.contributingModule()
+          .map(Element::asType)
+          .map(ComponentRequirement::forModule)
+          .map(module -> componentRequirementExpressions.getExpression(module, requestingClass));
+    } else if (compilerOptions.kotlinMetadata() && provisionBinding.isModuleKotlinObject()) {
+      return Optional.of(CodeBlock.of("$T.INSTANCE", provisionBinding.bindingTypeElement().get()));
```

Comment:
Confirming: is `.INSTANCE` always the name here?

Reply:
As far as I know, yes (per the docs - https://kotlinlang.org/docs/reference/java-to-kotlin-interop.html)

Reply:
What if there's a named companion object? Isn't that allowed?

Reply:
Yes, but this isn't affecting companion objects right now. Just top level objects